ci/scripts/check_dv_sw_images.py

```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python3
+
+# Copyright lowRISC contributors (OpenTitan project).
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+"""Checks for Bazel targets in sim DV cfg files.
+"""
+
+import sys
+import hjson
+from pathlib import Path
+
+from lib.bazel_query import BazelQueryRunner
+
+# List of fake targets added to please the linker:
+FAKE_TARGETS = [
+    "@manufacturer_test_hooks//:example_test",
+]
+
+
+def check_sw_image(name, sw_image, valid_sw_images):
+    # The format is target:type[:flags]...
+    # where the target is a bazel name so it also contains a ':'
+    # and the type indicate whether it's a flash image, otp image, and so on.
+    try:
+        bazel_module, bazel_label, other = sw_image.split(':', 2)
+    except ValueError:
+        print(f"- {name}: invalid sw_image '{sw_image}'")
+        return False
+    bazel_target = f'{bazel_module}:{bazel_label}'
+    if bazel_target not in valid_sw_images:
+        print(f"- {name}: non-existent target '{bazel_target}'")
+        return False
+    return True
+
+
+def check_links(cfg_hjson, valid_sw_images):
+    sim_cfg = hjson.load(Path(cfg_hjson).open())
+    ok = True
+    if 'tests' not in sim_cfg:
+        print("{} looks invalid: there are no tests")
+        return False
+    for test in sim_cfg['tests']:
+        name = test.get('name', '<invalid test name>')
+        sw_images = test.get('sw_images', [])
+        for sw_image in sw_images:
+            ok = check_sw_image(name, sw_image, valid_sw_images) and ok
+    return ok
+
+
+def usage():
+    print("usage: check_dv_sw_image.py [hjson cfg files]...")
+
+
+if __name__ == '__main__':
+    bazel = BazelQueryRunner()
+
+    # Search for all targets in device software.
+    valid_sw_images = set(bazel.query("//sw/device/..."))
+    valid_sw_images |= set(FAKE_TARGETS)
+    print("{} sim_dv targets identified".format(len(valid_sw_images)))
+
+    # Verify all files passed as argument
+    if len(sys.argv) == 1:
+        usage()
+        sys.exit(1)
+
+    ok = True
+    for f in sys.argv[1:]:
```

Comment:
I don't massively like this. I'd probably suggest making the Python script take a single hjson file and putting the loop into `check_dv_sw_images.sh`.

Reply:
One reason to do it this way is that potentially query the list of all bazel targets is an expensive operation. By doing things this way, we can factor the cost and really processing one or ten files does not make a difference in code complexity.

Reply:
Oh, good point. In which case, this makes loads of sense. Thanks for explaining.